action/sealedenvelope.go

```diff
@@ -48,27 +52,46 @@ func (sealed *SealedEnvelope) envelopeHash() (hash.Hash256, error) {
 // Hash returns the hash value of SealedEnvelope.
 // an all-0 return value means the transaction is invalid
 func (sealed *SealedEnvelope) Hash() (hash.Hash256, error) {
+	if sealed.hash != hash.ZeroHash256 || sealed.hashErr != nil {
+		return sealed.hash, sealed.hashErr
+	}
+	sealed.calcHash()
+	return sealed.hash, sealed.hashErr
+}
+
+func (sealed *SealedEnvelope) calcHash() {
```

Comment:
one final comment:
```
func (sealed *SealedEnvelope) calcHash() (hash.Hash256, error)
```
then don't need to change this func at all, in the above, just
```
sealed.hash, sealed.hashErr = calcHash()
```